client/src/main/scala/org/http4s/client/websocket/WSClient.scala

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.websocket
+
+import cats._
+import cats.data.Chain
+import cats.data.OptionT
+import cats.effect._
+import cats.effect.concurrent.Deferred
+import cats.effect.concurrent.Ref
+import cats.effect.concurrent.TryableDeferred
+import cats.implicits._
+import fs2.Pipe
+import fs2.Stream
+import org.http4s.Headers
+import org.http4s.Method
+import org.http4s.Uri
+import scodec.bits.ByteVector
+
+/** A websocket request.
+  *
+  * @param uri
+  *   The URI.
+  * @param headers
+  *   The headers to send. Put your `Sec-Websocket-Protocol` headers here if needed. Some websocket
+  *   clients reject other WS-specific headers.
+  * @param method
+  *   The method of the intial HTTP request. Ignored by some clients.
+  */
+final case class WSRequest(
```

Comment:
Should we mask the constructor for this like we do with some of our other case classes, e.g. [RequestPrelude](https://github.com/http4s/http4s/blob/2abafd2b5b265a28b27a1c8d2cbd05dd40ec4f9e/core/shared/src/main/scala/org/http4s/RequestPrelude.scala#L31)

Reply:
```suggestion
final case class WSRequest private (
```

Reply:
That won't really help bincompat with the unapply method still leaking, right?

Reply:
Yeah, and the `copy` methods. I can de-case-class this, that's always the safest bet.

Reply:
I personally like `private[this]` case classes in the companion object of a `sealed trait` like is done in RequestPrelude. It still lets the library authors access the copy methods and stuff.

Reply:
Yeah, that was a helpful pointer to `RequestPrelude`. I liked it too so I copied the same pattern here.